### ISSUE TITLE
Optimize weighted embedding extraction to do only 1 forward pass of the audio instead of `num_speakers`

### ISSUE THREAD

**Issue by juanmc2005** [juanmc2005/diart#214]

**Comment by juanmc2005** [comment]
After a quick run of the diarization pipeline with 5s latency, it seems that pyannote 3.1 is hurting performance: DER=27.3 -> DER=29.1 with the exact same models and hyper-parameters.

This is most certainly due to the new API for embedding models compatible with weighted stats pooling, so contrary to my initial idea of leaving this for v0.10, I think it's more a compatibility issue and less a matter of latency optimization.

**Comment by juanmc2005** [comment]
cc @hbredin, in case you can think of any other changes that could have caused this

**Comment by hbredin** [comment]
> cc @hbredin, in case you can think of any other changes that could have caused this

I can't think of any. I did not witness any change of performance in offline speaker diarization when switching from `hbredin/wespeaker-voxceleb-resnet34-LM` (ONNX) to `pyannote/wespeaker-voxceleb-resnet34-LM`(PyTorch).

**Comment by juanmc2005** [comment]
@hbredin I narrowed it down to a change in the interpolation method for the weights:

```diff
- weights = F.interpolate(
-     weights, size=num_frames, mode="linear", align_corners=False
- )
+ weights = F.interpolate(weights, size=num_frames, mode="nearest")
```

Any particular reason for this change? I guess interpolating before calling the model should do the trick but I'm curious.

**Comment by hbredin** [comment]
MPS support.

**Comment by juanmc2005** [comment]
@hbredin MPS is needed for training though, right? Any way we can make it apply a different strategy? Because to interpolate before calling the model I actually need to know the number of frames that the model will produce internally right before the stats pooling.

I suggest something like:
```python
model = Model.from_pretrained("pyannote/embedding")
model.set_interpolation_method("linear")
embeddings = model(waveform, weights)
```

Inside the `forward` method of the model:
```python
outputs = self.stats_pool(outputs, weights=weights, method=self.interpolation_method)
```

And to call the interpolation:
```python
def forward(..., interpolation_method: str = "nearest"):
    ...
    weights = F.interpolate(weights, size=num_frames, mode=interpolation_method)
    ...
```

Notice that the interpolation method would need to be changed with a setter because `StatsPool` is (as it should) created during instantiation.

Also, the same thing could be implemented for the masks in other embedding models (i.e. speechbrain and nemo). Right now I can't take advantage of this optimization because I would need to keep track of which model is compatible with it, and the rule may very easily change in the future, leading to compatibility hell.

I'm willing to open PRs for both features.

**Comment by juanmc2005** [comment]
Since everything is working fine except for that performance loss on AMI, I think I'd prefer not to wait for this fix to release v0.9. I'll add a note to the reproducibility section of the README to indicate that pyannote<3.1 should be used and link this issue.

**Comment by shenberg** [comment]
Hi, I'm raising this issue to note that as far as I can tell, the missing functionality has been implemented in the `mps` backend since PyTorch 2.3.0 ([The specific commit](https://github.com/pytorch/pytorch/commit/c59b14163b9f8ffc55f3dd5c155dbbd1535d9d80))

On my M1, using PyTorch 2.4.0:

```
>>> a = torch.arange(10, device='mps')[None, None].float() # size 1x1x10
>>> F.interpolate(a, size=20, mode='linear', align_corners=False)
tensor([[[0.0000, 0.2500, 0.7500, 1.2500, 1.7500, 2.2500, 2.7500, 3.2500,
          3.7500, 4.2500, 4.7500, 5.2500, 5.7500, 6.2500, 6.7500, 7.2500,
          7.7500, 8.2500, 8.7500, 9.0000]]], device='mps:0')
>>> F.interpolate(a, size=5, mode='linear', align_corners=False)
tensor([[[0.5000, 2.5000, 4.5000, 6.5000, 8.5000]]], device='mps:0')
```

so there should be no obstacle to returning to the original behavior, right?

**Comment by juanmc2005** [comment]
@shenberg have you checked whether using `mode="linear"` and `mode="mps"` give the same results on AMI?
